src/ai/util/areaCodeFromZip.util.ts

```diff
@@ -0,0 +1,178 @@
+      )
+    }
+
+    this.logger.debug(
+      `Area codes for zip code ${normalizedZipCode} not found in cache, calling OpenAI`,
+    )
+    const areaCodes = await this.getAreaCodesFromOpenAI(normalizedZipCode)
+
+    if (areaCodes && areaCodes.length > 0) {
+      await this.saveAreaCodesToS3(normalizedZipCode, areaCodes)
+      return areaCodes
+    }
+
+    return null
+  }
+
+  private isCacheExpired(lastFetchedAt: string): boolean {
+    const lastFetched = new Date(lastFetchedAt)
+    const now = new Date()
+
+    return differenceInYears(now, lastFetched) >= CACHE_EXPIRY_YEARS
+  }
+
+  private async getAreaCodesFromS3(
+    zipCode: string,
+  ): Promise<ZipToAreaCodeEntry | null> {
+    try {
+      const bucket = process.env.ZIP_TO_AREA_CODE_BUCKET
+      if (!bucket) {
+        throw new Error(
+          'ZIP_TO_AREA_CODE_BUCKET environment variable is required',
+        )
+      }
+      const key = this.s3Service.buildKey(undefined, ZIP_TO_AREA_CODE_FILE)
+      const json = await this.s3Service.getFile(bucket, key)
+
+      if (!json) {
+        return null
+      }
+
+      const mappings = JSON.parse(json) as ZipToAreaCodeMapping
+      return mappings[zipCode] || null
+    } catch (error) {
+      this.logger.error(`Error getting area code mappings from S3: ${error}`)
+      return null
+    }
+  }
+
+  private async getAreaCodesFromOpenAI(
+    zipCode: string,
+  ): Promise<string[] | null> {
+    const prompt = `What are the area codes (NPA codes) for the zip code ${zipCode} in the United States?
+      Please respond with ONLY a JSON array of ara code strings (3-digit numbers), for example: ["415", "510].
+      If you cannot determin the ara codes, respond with an empty array: [].`
+
+    try {
+      const response = await this.aiService.llmChatCompletion(
+        [
+          {
+            role: 'user',
+            content: prompt,
+          },
+        ],
+        100,
+        0.1,
+        0.1,
+      )
+
+      if (!response?.content || typeof response.content !== 'string') {
+        return null
+      }
+
+      const content = response.content.trim()
+      const jsonContent = content
+        .replace(/\n?/g, '')
+        .replace(/```\n?/g, '')
+        .trim()
+
+      const areaCodes = JSON.parse(jsonContent) as unknown
+
+      if (
+        Array.isArray(areaCodes) &&
+        areaCodes.every((code) => typeof code === 'string')
+      ) {
+        return areaCodes as string[]
+      }
+      return null
+    } catch (error) {
+      this.logger.error(
+        `Error getting area codes from OpenAI for zip ${zipCode}: ${error}`,
+      )
+      return null
+    }
+  }
+
+  private async saveAreaCodesToS3(
+    zipCode: string,
+    areaCodes: string[],
+  ): Promise<void> {
+    try {
+      const bucket = process.env.ZIP_TO_AREA_CODE_BUCKET
+      if (!bucket) {
+        throw new Error(
+          'ZIP_TO_AREA_CODE_BUCKET environment variable is required',
+        )
+      }
```

Comment:
Yeah.. moving this out to the top of the file will eliminate for this duplicative code